# Viral Genomics and Bioinformatics Latin America and the Caribbean LAC 2022
Base for the Viral Genomics and Bioinformatics Repository

## Course overview
We are pleased to announce the next Viral Genomics and Bioinformatics course in collaboration with COVID-19 Genomics (COG) Train (COG-Train).

Viruses are a huge public health burden all around the world.  Latin America is bearing the burden of the SARS-CoV-2 virus in addition to endemic viruses such as zika, chikungunya and dengue. Apart from health issues, emerging viral illnesses and epidemics have a global economic impact. According to the World Bank, the SARS-Cov-2 pandemic caused a more than $10 trillion deficit in the global economy in 2020-2021 alone. This demonstrates the impact of viral illnesses not only on human and animal health, but also on social and economic growth of the world.

Technological advances in science and research are making it possible to rapid detection, surveillance and management of viral diseases. In recent years, next generation sequencing (NGS) technologies have played an important role in the identification and classification of viruses, in the detection of drug resistance mutations and in the treatment and surveillance of viral diseases. Early identification of a virus and quick analysis of its genome will aid towards better treatment and help in controlling the disease spread. Along with the advances in the sequencing technologies, it is necessary to develop data management methods, computational analysis tools and human expertise to cope with the sequence deluge. The number of researchers able to interpret viral high throughput sequence data is currently limited.   

The Viral Genomics and Bioinformatics course will teach participants how to analyse  viral genome sequences and interpret genomic data obtained by high-throughput sequencing. The use of the command line to analyse viral NGS data, quality control of sequences, reference mapping and de novo assembly of viral genomes, pathogen identification from metagenomics data, and creating phylogenetic trees will be covered in the course. Participants, along with gaining bioinformatics knowledge, will have an opportunity to establish links and networks and develop future collaborative projects.

[Course website](https://coursesandconferences.wellcomeconnectingscience.org/event/viral-genomics-and-bioinformatics-latin-america-and-the-caribbean-20221010/)

## Instructors
- [Carolina Torres](https://www.researchgate.net/profile/Carolina-Torres-7), Universidad de Buenos Aires - CONICET, Buenos Aires, Argentina     
- [Josefina Campos](https://www.researchgate.net/profile/Josefina-Campos-2), Centro Nacional de Genómica y Bioinfomática, ANLIS Malbrán, Argentina
- [Sreenu Vattipally](https://www.gla.ac.uk/researchinstitutes/iii/staff/sreenuvattipally/), MRC-University of Glasgow Centre for Virus Research, UK
- [Marta Giovanetti](), Oswaldo Cruz Foundation, Brazil
- [Pablo Ramos](https://cidacs.bahia.fiocruz.br/en/), FIOCRUZ - Oswaldo Cruz Foundation, Brazil
- [Richard Orton](https://www.gla.ac.uk/researchinstitutes/iii/staff/richardorton/), MRC-University of Glasgow Centre for Virus Research, UK
- [Srikeerthana Kuchi](), MRC-University of Glasgow Centre for Virus Research, UK
- [Maria Dolores Blanco Fernandez](Link), Faculty of Pharmacy and Biochemistry, University of Buenos Aires, Argentina
- [András Carlos Alberto Culasso ](Link), CONICET - Faculty of Pharmacy and Biochemistry, University of Buenos Aires, Argentina     
- [Laise de Moraes](https://www.researchgate.net/profile/Laise-De-Moraes), FIOCRUZ-BA, Brazil
- [Tomás Poklépovich](https://www.researchgate.net/profile/Tomas-Poklepovich), Centro Nacional de Genómica y Bioinfomática, ANLIS Malbrán, Argentina
- [María Sol Haim](Link), Centro Nacional de Genómica y Bioinfomática, ANLIS Malbrán, Argentina







## Timetable
### Overview


### Detailed timetable
[View Timetable here](https://docs.google.com/document/d/1sBF9T7hCbNgtoI3THcaht030gFDwiiuY/edit?usp=sharing&ouid=101020480766226270980&rtpof=true&sd=true)

### Advanced Learning and Training Handout
[View ALT Handout here](Modules/Advanced%20Courses%20ALT%20Sessions%20handouts%20edited.pdf)

## Course manual
**Module 1 - Introduction to Linux + Module 2 - Linux text processing**   
[PDF Manual - Introduction to Linux + Text processing](Modules/Intro_to_linux_text_processing.pdf)     
<!-- [PDF Lectures - Introduction to Linux + Text processing](Modules/Lectures_and_Answers/Intro_to_linux_Asia2022.pdf)  -->
<!-- [PDF Solutions - Introduction to Linux + Text processing](Modules/Lectures_and_Answers/Linux_solutions_to_exercises_and_quiz.pdf) -->

**Module 3 - Introduction to NGS sequencing**  
**Marta + Xoilson - check **
<!-- [Online Manual - Introduction to NGS](Modules/Introduction_to_NGS.md) -->
  
**Module 4 - NGS file formats and data & QC**  
[Online Manual - NGS data Quality Control QC](Modules/NGS_file_formats_and_QC.md)

**Module 5 - Introduction to BASH scripting**  
** Josefina -  check **
[PDF Manual - Bash Scripting](Modules/BASH_scripting.pdf)      
<!-- [PDF Answers - Bash Scripting](Modules/Lectures_and_Answers/bash_scripting/code_description_BASH-Practical.pdf).  -->
<!-- [Answer Scripts Folder - Bash Scripting](Modules/Lectures_and_Answers/bash_scripting/)  -->


**Module 6 - Reference Alignment**  
[Online Manual - Reference Alignment](Modules/Reference_alignment.md)
 
**Module 7 - Coverage Plots and Stats**  
[Online Manual - Coverage Plots and stats](Modules/Coverage_plots_and_statistics.md)
 
**Module 8 - Consensus & variant calling**  
[Online Manual - Consensus & variant calling](Modules/Consensus-and-variant-calling-theory-and-practical.md)

**Module 9 - De Novo assembly**  
[PDF Manual - De Novo assembly](Modules/DeNovo-Sreenu-Asia-2022.pdf)      

**Module 10 - Pathogen sequence detection using metagenomics**  
[Online Manual - Metagenomics](course_data/Pathogen_sequence_detection_using_metagenomics/Module10-Metagnomics.md)

[PDF Manual - Metagenomics](course_data/Pathogen_sequence_detection_using_metagenomics/Module10-Metagnomics.pdf)


**Module 11 - Introduction to phylogenetic methods**  

[Online Manual - Phylogenetics and Virus Typing](Modules/Phylogeny_&_Virus_Typing_HandsOnEx.md)   
[PDF Phylogenetics_Hands-on exercises](Modules/Phylogenetics_Hands-on_exercises.pdf)    

**Module 12 - SARS-CoV-2 workflows**  

[Online Manual - SARS-CoV-2 Reference Alignment](https://github.com/WCSCourses/ViralBioinfAsia2022/blob/main/Modules/SARS-CoV-2.md)

[Online Manual - SARS-CoV-2 Lineages and Mutations](https://github.com/WCSCourses/ViralBioinfAsia2022/blob/main/Modules/SARS-CoV-2.md#2-sars-cov-2-lineages-and-mutations)

[Online Manual - SARS-CoV-2 Phylogenetics](https://github.com/WCSCourses/ViralBioinfAsia2022/blob/main/Modules/SARS-CoV-2.md#3-sars-cov-2-phylogenetics)

[Online Manual - SARS-CoV-2 Group Practical](https://github.com/WCSCourses/ViralBioinfAsia2022/blob/main/Modules/SARS-CoV-2.md#4-sars-cov-2-group-practical) 



**Appendix**      
<!-- [PDF - ALT Learning Outcomes Session 2 Slides](Modules/COG-TrainALTslides-Session2.pdf)    + -->
<!-- [PDF - Answers ALT Learning Outcomes Exercise](https://github.com/WCSCourses/ViralBioinfAsia2022/blob/766ad82cc5b841448adda9377e977363668ba3a9/Modules/9_ALT_LO_answersCOG-Train%20ALT%20slides%20-%20Session2.pdf)       + -->
      



Any reuse of the course materials, data or code is encouraged with due acknowledgement.

******
## License
<a rel="license" href="http://creativecommons.org/licenses/by/4.0/"><img alt="Creative Commons Licence" style="border-width:0" src="https://i.creativecommons.org/l/by/4.0/88x31.png" /></a><br />This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution 4.0 International License</a>.

